### PR TITLE
Fix variance calculation for subtotal differences

### PR DESCRIPTION
**Pull request by gergness**
We didn't update the proportion variance calculation to account for subtotal differences. See https://www.genroe.com/blog/how-to-calculate-margin-of-error-and-other-stats-for-nps/5994 for more information about the formula.

### CODE REVIEW

File: src/cr/cube/matrix/measure.py [hunk 1]
```diff
@@ -2310,10 +2339,11 @@ def _counts(self):
             counts = self._second_order_measures.column_comparable_counts.blocks
             # --- Get base values & *row* subtotals
             return [counts[0][0], counts[1][0]]
-        # --- Use *row* comparable counts (going across columns) ---
-        counts = self._second_order_measures.row_comparable_counts.blocks
-        # --- Get base values & *column* subtotals
-        return [counts[0][0], counts[0][1]]
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2342]: i think this else is not needed

File: src/cr/cube/matrix/measure.py [hunk 2]
```diff
@@ -2391,10 +2421,14 @@ def _proportion_numerators(self):
         counts = self._second_order_measures.weighted_counts.blocks
 
         if self.orientation == MO.ROWS:
-            # --- Now get only the base values and relevant row subtotals
+            # --- Try getting the column comparable counts, to check for the error
+            self._second_order_measures.column_comparable_counts.blocks
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2425]: mmm this call doesn't do anything 🤷🏽‍♂️ 
it seems weird to me checking errors just calling a property

File: src/cr/cube/matrix/measure.py [hunk 3]
```diff
@@ -2391,10 +2421,14 @@ def _proportion_numerators(self):
...
+            # --- Now get only the base values and relevant subtotals
             counts = [counts[0][0], counts[1][0]]
         else:
-            # --- Now get only the base values and relevant column subtotals
+            # --- Try getting the row comparable counts, to check for the error
+            self._second_order_measures.row_comparable_counts
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2430]: same

File: src/cr/cube/matrix/measure.py [hunk 4]
```diff
@@ -2404,7 +2438,8 @@ def _proportion_denominators(self):
         """list of ndarray 1D table margins to be used as denominator"""
         if self.orientation == MO.ROWS:
             return self._second_order_measures.rows_table_weighted_base.blocks
-        return self._second_order_measures.columns_table_weighted_base.blocks
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2441]: this else is not needed

File: src/cr/cube/matrix/measure.py [hunk 5]
```diff
@@ -2504,8 +2539,9 @@ def blocks(self):
         if self.orientation == MO.ROWS:
             bases = self._second_order_measures.row_unweighted_bases.blocks
             return [bases[0][0][:, 0], bases[1][0][:, 0]]
-        bases = self._second_order_measures.column_unweighted_bases.blocks
-        return [bases[0][0][0, :], bases[0][1][0, :]]
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2542]: same

File: src/cr/cube/matrix/measure.py [hunk 6]
```diff
@@ -2541,8 +2577,9 @@ def blocks(self):
         if self.orientation == MO.ROWS:
             bases = self._second_order_measures.row_weighted_bases.blocks
             return [bases[0][0][:, 0], bases[1][0][:, 0]]
-        bases = self._second_order_measures.column_weighted_bases.blocks
-        return [bases[0][0][0, :], bases[0][1][0, :]]
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2580]: same

File: src/cr/cube/matrix/measure.py [hunk 7]
```diff
@@ -2607,10 +2644,11 @@ def _proportions(self):
             props = self._second_order_measures.row_proportions.blocks
             # --- Get base values & *row* subtotals
             return [props[0][0], props[1][0]]
-        # --- Use *column* proportions ---
-        props = self._second_order_measures.column_proportions.blocks
-        # --- Get base values & *column* subtotals
-        return [props[0][0], props[0][1]]
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2647]: same

File: src/cr/cube/matrix/measure.py [hunk 8]
```diff
@@ -2732,7 +2770,8 @@ def _weighted_median(sorted_counts, sorted_values):
         # --- TODO: Does this behave okay for large numbers (or is precision a problem)?
         if cumulative_prop[median_idx] == 0.5:
             return np.mean(sorted_values[[median_idx, median_idx + 1]])
-        return sorted_values[median_idx]
+        else:
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:2773]: did you change all this stuff for some reason? Sorry just out of curiosity maybe we decided something and I forgot

File: tests/unit/matrix/test_measure.py [hunk 9]
```diff
@@ -127,6 +123,71 @@ def it_provides_access_to_various_measure_objects(
         MeasureCls_.assert_called_once_with(dimensions_, measures, cube_measures_)
         assert measure is measure_
 
+    @pytest.mark.parametrize(
+        "measure_prop_name, MeasureCls, proportion_type, base_type",
+        (
+            (
+                "column_proportion_variances",
+                _ProportionVariances,
```
Comment (ernestoarbitrio) [inline tests/unit/matrix/test_measure.py:131]: I think the `MeasureCls` should not be a param of the test considering that it's always `_ProportionVariances`

**Review by ernestoarbitrio — APPROVED** [review]

**Review by ernestoarbitrio — CHANGES_REQUESTED** [review]
per comments

File: src/cr/cube/matrix/measure.py (hunk 1 above)
Comment (gergness) [reply src/cr/cube/matrix/measure.py:2342]: Shoot, I must have started from a file that had already been edited, I didn't mean to change this line. Will investigate what happened

Comment (gergness) [reply src/cr/cube/matrix/measure.py]: Ok yea, must have started from an edited file, I did a rebase so I could make sure I only changed where I thought.

File: src/cr/cube/matrix/measure.py [hunk 10]
```diff
@@ -790,37 +810,110 @@ def _subtotal_rows(self):
...
-                p[1][0] * (1 - p[1][0]),
+                self._calc_var(p[1][0], Nt[1][0], Np[1][0], Ni[1][0], Nn[1][0]),
                 # --- intersections ---
-                p[1][1] * (1 - p[1][1]),
+                self._calc_var(p[1][1], Nt[1][1], Np[1][1], Ni[1][1], Nn[1][1]),
+            ],
+        ]
+
+    def _calc_var(self, p, Nt, Np, Ni, Nn):
+        """2D ndarray of float64 representing variance.
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/measure.py:870]: This is a function/method so the docstring should say:
```Returns a float64 2D ndarray representing variance.```

File: src/cr/cube/matrix/subtotals.py [hunk 11]
```diff
@@ -149,6 +149,81 @@ def _subtotal_row(self, subtotal):
...
+class PositiveTermSubtotals(_BaseSubtotals):
+    """Subtotal blocks that ignore the negative terms from differences
+
+    These values are the sum of the "positive" terms (the category for regular
+    categories, the terms in a regular subtotal, and the sum of just the positive terms
+    in a subtotal difference).
+    """
+
+    def _intersection(self, row_subtotal, column_subtotal):
+        """Positive terms from row/column subtotal intersection."""
```
Comment (ernestoarbitrio) [inline src/cr/cube/matrix/subtotals.py:208]: fix docstring here cause it's a method so it should say .. .Returns positive etc.....

**Review by ernestoarbitrio — APPROVED** [review]
LGTM. Just fix the 2 dosctrings I mentioned with the comments

**Pull request merged**